Eliminate exception throwing code from dispatch call sites

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32168 Eliminate exception throwing code from dispatch call sites**
* #32121 Out-of-line construction of OperatorName.
* #32118 Move error reporting code out-of-line from header.
* #32117 Make an assert on a hotpath trigger only in DEBUG mode.

We move the exception raising into the function, saving us a
big pile of instructions for raising the stack.

After this stack of changes, the compiler is willing to inline, e.g.,
`c10::KernelFunction::callUnboxed<at::Tensor, at::Tensor const&>(c10::OperatorHandle const&, at::Tensor const&) const::__func__`
(whereas previously it refused to do so.)

Taking this and the previous three commits in net, we see the following improvement according to cachegrind:

Before:

```
==2063141== I refs: 4,598,186,943
==2063141== I1 misses: 29,915,243
==2063141== LLi misses: 141,170
==2063141== I1 miss rate: 0.65%
==2063141== LLi miss rate: 0.00%
==2063141== 
==2063141== D refs: 2,346,841,256 (1,395,422,290 rd + 951,418,966 wr)
==2063141== D1 misses: 1,667,522 ( 1,039,423 rd + 628,099 wr)
==2063141== LLd misses: 510,723 ( 205,257 rd + 305,466 wr)
==2063141== D1 miss rate: 0.1% ( 0.1% + 0.1% )
==2063141== LLd miss rate: 0.0% ( 0.0% + 0.0% )
==2063141== 
==2063141== LL refs: 31,582,765 ( 30,954,666 rd + 628,099 wr)
==2063141== LL misses: 651,893 ( 346,427 rd + 305,466 wr)
==2063141== LL miss rate: 0.0% ( 0.0% + 0.0% )
```

after

```
==2090999== 
==2090999== I refs: 4,671,555,352
==2090999== I1 misses: 11,856,396
==2090999== LLi misses: 141,779
==2090999== I1 miss rate: 0.25%
==2090999== LLi miss rate: 0.00%
==2090999== 
==2090999== D refs: 2,400,048,122 (1,431,551,726 rd + 968,496,396 wr)
==2090999== D1 misses: 1,672,599 ( 1,044,558 rd + 628,041 wr)
==2090999== LLd misses: 509,699 ( 204,191 rd + 305,508 wr)
==2090999== D1 miss rate: 0.1% ( 0.1% + 0.1% )
==2090999== LLd miss rate: 0.0% ( 0.0% + 0.0% )
==2090999== 
==2090999== LL refs: 13,528,995 ( 12,900,954 rd + 628,041 wr)
==2090999== LL misses: 651,478 ( 345,970 rd + 305,508 wr)
==2090999== LL miss rate: 0.0% ( 0.0% + 0.0% )
```

Notice that the I1 misses get HALVED.

This is on ngimel's C++ benchmark; I built it with `buck build @mode/opt //experimental/ngimel/benchmark_framework_overheads:cpp_benchmark`

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392948](https://our.internmc.facebook.com/intern/diff/D19392948)

ezyang committed:
Eliminate exception throwing code from dispatch call sites

We move the exception raising into the function, saving us a
big pile of instructions for raising the stack.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 7d3324df:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7d3324df516af2b55e906542260dc677a8f750a1).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4237041) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=85795161) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=85795161))
<details>
<summary>
<code>Jan 14 15:15:25 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 14 15:15:25 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 14 15:15:25 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 14 15:15:25 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 14 15:15:25 1 out of 1 hunk FAILED 
Jan 14 15:15:25 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 14 15:15:25 Traceback (most recent call last): 
Jan 14 15:15:25   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 14 15:15:25     patch_repo(args) 
Jan 14 15:15:25   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 14 15:15:25     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 14 15:15:25 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 14 15:15:25 + cleanup 
Jan 14 15:15:25 + retcode=1 
Jan 14 15:15:25 + set +x 
Jan 14 15:15:25 =================== sccache compilation log =================== 
Jan 14 15:15:25 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 14 15:15:25 Compile requests                 1 
Jan 14 15:15:25 Compile requests executed        0 
Jan 14 15:15:25 Cache hits                       0 
Jan 14 15:15:25 Cache misses                     0 
Jan 14 15:15:25 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ezyang committed:
Update on "Eliminate exception throwing code from dispatch call sites"


We move the exception raising into the function, saving us a
big pile of instructions for raising the stack.

After this stack of changes, the compiler is willing to inline, e.g.,
`c10::KernelFunction::callUnboxed<at::Tensor, at::Tensor const&>(c10::OperatorHandle const&, at::Tensor const&) const::__func__`
(whereas previously it refused to do so.)

Taking this and the previous three commits in net, we see the following improvement according to cachegrind:

Before:

```
==2063141== I refs: 4,598,186,943
==2063141== I1 misses: 29,915,243
==2063141== LLi misses: 141,170
==2063141== I1 miss rate: 0.65%
==2063141== LLi miss rate: 0.00%
==2063141== 
==2063141== D refs: 2,346,841,256 (1,395,422,290 rd + 951,418,966 wr)
==2063141== D1 misses: 1,667,522 ( 1,039,423 rd + 628,099 wr)
==2063141== LLd misses: 510,723 ( 205,257 rd + 305,466 wr)
==2063141== D1 miss rate: 0.1% ( 0.1% + 0.1% )
==2063141== LLd miss rate: 0.0% ( 0.0% + 0.0% )
==2063141== 
==2063141== LL refs: 31,582,765 ( 30,954,666 rd + 628,099 wr)
==2063141== LL misses: 651,893 ( 346,427 rd + 305,466 wr)
==2063141== LL miss rate: 0.0% ( 0.0% + 0.0% )
```

after

```
==2090999== 
==2090999== I refs: 4,671,555,352
==2090999== I1 misses: 11,856,396
==2090999== LLi misses: 141,779
==2090999== I1 miss rate: 0.25%
==2090999== LLi miss rate: 0.00%
==2090999== 
==2090999== D refs: 2,400,048,122 (1,431,551,726 rd + 968,496,396 wr)
==2090999== D1 misses: 1,672,599 ( 1,044,558 rd + 628,041 wr)
==2090999== LLd misses: 509,699 ( 204,191 rd + 305,508 wr)
==2090999== D1 miss rate: 0.1% ( 0.1% + 0.1% )
==2090999== LLd miss rate: 0.0% ( 0.0% + 0.0% )
==2090999== 
==2090999== LL refs: 13,528,995 ( 12,900,954 rd + 628,041 wr)
==2090999== LL misses: 651,478 ( 345,970 rd + 305,508 wr)
==2090999== LL miss rate: 0.0% ( 0.0% + 0.0% )
```

Notice that the I1 misses get HALVED.

This is on ngimel's C++ benchmark; I built it with `buck build @mode/opt //experimental/ngimel/benchmark_framework_overheads:cpp_benchmark`

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392948](https://our.internmc.facebook.com/intern/diff/D19392948)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -73,10 +73,19 @@ class CAFFE2_API Dispatcher final {
+   * it raises an exception if the operator is not found (so we skip
+   * generating exception raising code at the call site)
+   *
+   * Irritatingly, we still have to generate the handful of instructions
```
did you test returning `c10::optional` and then making it `noexcept`?

ezyang replied:
No, but I know what it looks like: you skip the guard failure handling (a few instructions) and bring back exception raising (a lot of instructions).

ezyang replied:
(or maybe you don't even get to skip guard failure, that seems likely to me)

ezyang committed:
Update on "Eliminate exception throwing code from dispatch call sites"


We move the exception raising into the function, saving us a
big pile of instructions for raising the stack.

After this stack of changes, the compiler is willing to inline, e.g.,
`c10::KernelFunction::callUnboxed<at::Tensor, at::Tensor const&>(c10::OperatorHandle const&, at::Tensor const&) const::__func__`
(whereas previously it refused to do so.)

Taking this and the previous three commits in net, we see the following improvement according to cachegrind:

Before:

```
==2063141== I refs: 4,598,186,943
==2063141== I1 misses: 29,915,243
==2063141== LLi misses: 141,170
==2063141== I1 miss rate: 0.65%
==2063141== LLi miss rate: 0.00%
==2063141== 
==2063141== D refs: 2,346,841,256 (1,395,422,290 rd + 951,418,966 wr)
==2063141== D1 misses: 1,667,522 ( 1,039,423 rd + 628,099 wr)
==2063141== LLd misses: 510,723 ( 205,257 rd + 305,466 wr)
==2063141== D1 miss rate: 0.1% ( 0.1% + 0.1% )
==2063141== LLd miss rate: 0.0% ( 0.0% + 0.0% )
==2063141== 
==2063141== LL refs: 31,582,765 ( 30,954,666 rd + 628,099 wr)
==2063141== LL misses: 651,893 ( 346,427 rd + 305,466 wr)
==2063141== LL miss rate: 0.0% ( 0.0% + 0.0% )
```

after

```
==2090999== 
==2090999== I refs: 4,671,555,352
==2090999== I1 misses: 11,856,396
==2090999== LLi misses: 141,779
==2090999== I1 miss rate: 0.25%
==2090999== LLi miss rate: 0.00%
==2090999== 
==2090999== D refs: 2,400,048,122 (1,431,551,726 rd + 968,496,396 wr)
==2090999== D1 misses: 1,672,599 ( 1,044,558 rd + 628,041 wr)
==2090999== LLd misses: 509,699 ( 204,191 rd + 305,508 wr)
==2090999== D1 miss rate: 0.1% ( 0.1% + 0.1% )
==2090999== LLd miss rate: 0.0% ( 0.0% + 0.0% )
==2090999== 
==2090999== LL refs: 13,528,995 ( 12,900,954 rd + 628,041 wr)
==2090999== LL misses: 651,478 ( 345,970 rd + 305,508 wr)
==2090999== LL miss rate: 0.0% ( 0.0% + 0.0% )
```

Notice that the I1 misses get HALVED.

This is on ngimel's C++ benchmark; I built it with `buck build @mode/opt //experimental/ngimel/benchmark_framework_overheads:cpp_benchmark`

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392948](https://our.internmc.facebook.com/intern/diff/D19392948)

[ghstack-poisoned]
